### PR TITLE
Addition of the nlp quick tip on Explainability for Transformers using SHAP

### PR DESCRIPTION
**Pull request by VHeusinkveld**
Notebook + readme
- How to use SHAP in combination with Huggingface Transformers
- SHAP applied to 3 types of examples
- Interactive widget, for experimentation on collab
- GPU implementation in the appendix (20x performance compared to CPU)

### CODE REVIEW

File: nlp/2021_04_22_shap_for_huggingface_transformers/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# BERT, explain yourself! 📖
+
+It has been over two years since transformer models took the NLP throne 🏅, but up until recently they couldn't tell you how they got to their prediction. Not anymore!
+
+```python
+>>> BERT('After many years of training I finally succeeded in achieving '
```
Comment (ThomasVrancken) [inline nlp/2021_04_22_shap_for_huggingface_transformers/README.md:6]: Might be nice to use an example sentence about the post itself ? Smtg like "After many year of text classification SHAP now allows us to dig in the model's logic! This great tool... " 😃

**Comment by ThomasVrancken** [comment]
I couldn't open the notebook with the link, sure it's working ? I'm also using my personal address in Gitlab, so could be that it doesn't have access to the notebook yet. Maybe you can then just make it public (read only) already ?

**Comment by VHeusinkveld** [comment]
The issue is probably that the collab link already assumes we are on the main branch while we are on a side branch. This [link](https://colab.research.google.com/github/ml6team/quick-tips/blob/feature/nlp/shap-transformers/nlp/2021_04_22_shap_for_huggingface_transformers/explainable_transformers_using_shap.ipynb) should work. In addition, I shared the notebook privately with you for leaving comments.

**Comment by ThomasVrancken** [comment]
Nice one! Can surely say this will be useful. Actually I have a couple of questions:
1. Did you try it for multi-class classification ? 
2. Do you think it would be possible to use it with zero-shot classification model ? We're probably going to try using one at Signal in our next track, as we have a pretty simple problem and no labeled data. So using smtg like this: https://huggingface.co/facebook/bart-large-mnli

Though those questions are unrelated with the PR, so good to go imo ;-)

**Review by ThomasVrancken — APPROVED** [review]

File: nlp/2021_04_22_shap_for_huggingface_transformers/explainable_transformers_using_shap.ipynb [hunk 2]
```diff
@@ -0,0 +1,3700 @@
...
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "x31kAhFUap3Y"
+      },
+      "source": [
+        "As you might have guessed from the colors, BERT has learned to harness the power of [SHAP](https://shap.readthedocs.io/en/latest/index.html)! SHAP is a game theoretic approach to explain the output of any machine learning model. Darker shades of pink indicate a positive contribution and the shades of blue indicate a negative contribution to the sentiment. This enables us to get an understanding of what goes on under the hood and not take predictions at their face value.\n",
+        "\n",
+        "Let's dive 🤿 in and learn how you supercharge your NLP projects with explenations.\n",
```
Comment (TDehaene) [inline nlp/2021_04_22_shap_for_huggingface_transformers/explainable_transformers_using_shap.ipynb:552]: Small typo: explenations

**Review by TDehaene — APPROVED** [review]
Cool stuff :-)!
Apart from a small typo, don't have any further comments, thanks for the great contribution Vincenzo!

File: nlp/2021_04_22_shap_for_huggingface_transformers/README.md (hunk 1 above)
Comment (VHeusinkveld) [reply nlp/2021_04_22_shap_for_huggingface_transformers/README.md:6]: I think this is a nice suggestion, but given all the changes required I will leave it in the current example.

**Comment by VHeusinkveld** [comment]
> Nice one! Can surely say this will be useful. Actually I have a couple of questions:
> 
> 1. Did you try it for multi-class classification ?

This is directly transferable! Note that the problem in the example is phrased as Negative and Positive classification (i.e. already multi-class). A model with let's say 4 outputs will just work fine.

> 2. Do you think it would be possible to use it with zero-shot classification model ? We're probably going to try using one at Signal in our next track, as we have a pretty simple problem and no labeled data. So using smtg like this: https://huggingface.co/facebook/bart-large-mnli

This approach works if you have a problem where there is a constant number of classification outputs (hence should work). One thing that could be a problem: if it is a zero-shot problem with prompt engineering, the SHAP library will also mask out parts of the prompt to explain the output (with some tricks you can probably make sure that parts are never masked). Hope with this info you can make the assessment yourself:)   

> Though those questions are unrelated with the PR, so good to go imo ;-)
Thanks! Let's catch up in chat if something is still unclear, happy to help out.

File: nlp/2021_04_22_shap_for_huggingface_transformers/explainable_transformers_using_shap.ipynb (hunk 2 above)
Comment (VHeusinkveld) [reply nlp/2021_04_22_shap_for_huggingface_transformers/explainable_transformers_using_shap.ipynb]: Fixed:) nice spot

**Pull request merged**